src/Features/Core/Portable/SplitOrMergeIfStatements/AbstractMergeIfStatementsCodeRefactoringProvider.cs

```diff
@@ -0,0 +1,91 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeRefactorings;
+using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis.SplitOrMergeIfStatements
+{
+    internal abstract class AbstractMergeIfStatementsCodeRefactoringProvider<TExpressionSyntax> : CodeRefactoringProvider
+        where TExpressionSyntax : SyntaxNode
+    {
+        protected abstract string IfKeywordText { get; }
+
+        protected abstract bool IsApplicableSpan(SyntaxNode node, TextSpan span, out SyntaxNode ifStatementNode);
+
+        protected abstract bool IsIfStatement(SyntaxNode node);
+
+        protected abstract CodeAction CreateCodeAction(Func<CancellationToken, Task<Document>> createChangedDocument);
+
+        protected abstract Task<bool> CanBeMergedAsync(
+            Document document, SyntaxNode ifStatement, ISyntaxFactsService syntaxFacts, CancellationToken cancellationToken);
+
+        protected abstract SyntaxNode GetChangedRoot(
+            SyntaxNode root, SyntaxNode ifStatement, ISyntaxFactsService syntaxFacts, SyntaxGenerator generator);
+
+        public sealed override async Task ComputeRefactoringsAsync(CodeRefactoringContext context)
+        {
+            var root = await context.Document.GetSyntaxRootAsync(context.CancellationToken).ConfigureAwait(false);
+            var node = root.FindNode(context.Span, getInnermostNodeForTie: true);
+
+            if (IsApplicableSpan(node, context.Span, out var ifStatement))
+            {
+                var syntaxFacts = context.Document.GetLanguageService<ISyntaxFactsService>();
+
+                if (await CanBeMergedAsync(context.Document, ifStatement, syntaxFacts, context.CancellationToken).ConfigureAwait(false))
+                {
+                    context.RegisterRefactoring(
+                        CreateCodeAction(
+                            c => RefactorAsync(context.Document, context.Span, syntaxFacts, c)));
```

Comment:
nit.  teh sytaxFacts guy isn't really used in this method. a nd it's super cheap to acquire on demand.  it might be cleaner to just not pass along and have the called methods get it.

Reply:
I pushed it up to the base class as part of moving shared code and removing duplication, but I can move it back down.

Reply:
➡️ Fixed in dc77a01fe5f1e6fd0f96e94402ea7e9c5b8ee73c